### PR TITLE
Cartesian subplot fixes

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/1044

and fixes relayout problems discovered in http://community.plot.ly/t/resizing-is-not-working-after-setting-height-in-plotly-relayout/2196/6?u=etienne

These problems were introduced in https://github.com/plotly/plotly.js/pull/946

This PR will need to merge `master` after https://github.com/plotly/plotly.js/pull/1048 to :white_check_mark:

### CODE REVIEW

**Comment by etpinard** [comment]
@rreusser would mind taking a :eye: on this?

File: src/plots/plots.js [hunk 1]
```diff
@@ -402,6 +402,9 @@ plots.supplyDefaults = function(gd) {
     // clean subplots and other artifacts from previous plot calls
     plots.cleanPlot(newFullData, newFullLayout, oldFullData, oldFullLayout);
 
+    // relink / initialize subplot axis objects
+    plots.linkSubplots(newFullData, newFullLayout, oldFullData, oldFullLayout);
```
Comment (rreusser) [inline src/plots/plots.js]: Ah, so it was that the axes were outdated? That seems consistent with what was observed.

File: test/jasmine/tests/hover_label_test.js [hunk 2]
```diff
@@ -630,6 +631,16 @@ describe('hover after resizing', function() {
 
     afterEach(destroyGraphDiv);
 
+    function _click(pos) {
```
Comment (rreusser) [inline test/jasmine/tests/hover_label_test.js]: 👍  for fixing via testing!

File: test/jasmine/tests/plots_test.js [hunk 3]
```diff
@@ -248,7 +248,7 @@ describe('Test Plots', function() {
     describe('Plots.resize', function() {
         var gd;
 
-        beforeEach(function(done) {
+        beforeAll(function(done) {
```
Comment (rreusser) [inline test/jasmine/tests/plots_test.js]: Hmm… `beforeAll`? Usually when I've written that, it's not actually what I want. Does it leak state between the tests? Is it *definitely* what you meant here?

Comment (etpinard) [inline test/jasmine/tests/plots_test.js]: Yep, this is on purpose.

This particular `describe` creates a graph then calls `Plotly.Plots.resize`. The multiple `it` block are simply there to group the assertion in some logical way. There's no need to re-create the graph before each `it` in this case.

File: src/plots/plots.js (hunk 1 above)
Comment (etpinard) [inline src/plots/plots.js]: Exactly.

File: test/jasmine/tests/plots_test.js [hunk 4]
```diff
@@ -287,6 +287,17 @@ describe('Test Plots', function() {
...
+
+        it('should update the axis scales', function() {
+            var fullLayout = gd._fullLayout,
+                plotinfo = fullLayout._plots.xy;
+
+            expect(fullLayout.xaxis._length).toEqual(240);
+            expect(fullLayout.yaxis._length).toEqual(220);
+
+            expect(plotinfo.xaxis._length).toEqual(240);
+            expect(plotinfo.yaxis._length).toEqual(220);
```
Comment (etpinard) [inline test/jasmine/tests/plots_test.js]: In particular @rreusser , these assertions were failing before https://github.com/plotly/plotly.js/pull/1049/commits/f4ca110ac0b786eed371fa680cdd95c09d5225a2

Comment (rreusser) [inline test/jasmine/tests/plots_test.js]: Got it. Sounds good! I once misunderstood `beforeAll` to mean before each all tests, so thought it was worth checking!

**Comment by etpinard** [comment]
@rreusser 

> I'll definitely want to fire up a couple test cases once merged to confirm that this has fixed things

Agreed. I would be nice to beef up our resize test suites. Let's first publish `1.18.1`. 

Thanks very much for reviewing this!

**Comment by rreusser** [comment]
💃

**Pull request merged**